### PR TITLE
Downward-facing rangefinder based low altitude fence

### PR DESCRIPTION
**Pull request by amilcarlucas**
This fence prevents the copter from hitting the ground in all flight modes except STABALIZE, RTL and LAND.
If for some reason it does manage to go bellow the fence then RTL is triggered (RTL_ALT http://ardupilot.org/copter/docs/parameters.html#rtl-alt-rtl-altitude should be bigger than FENCE_ALT_MIN parameter).
RTL and LAND flight modes completely ignore this fence.

If no range finder is present it will use the height sensor selected in the EK2_ALT_SOURCE or EK3_ALT_SOURCE parameter ( http://ardupilot.org/copter/docs/parameters.html#ek2-alt-source-primary-altitude-sensor-source )

It contains CI tests.
laf_log

### CODE REVIEW

**Comment by magicrub** [comment]
>It contains CI tests.

bonus points! Although, its failing the CI test... :)

**Comment by amilcarlucas** [comment]
Unfortunatlly 5748866 is a big hack and the CI builds do not like that. I will have to fix it somehow.

**Comment by Katawann** [comment]
I tested the low fence which works as expected with SITL. But no correction happens when working with the AP_Proximity. I did a test:
- 4 WPs at 2 m of altitude
- FENCE_ALT_MIN = 2 m and FENCE_MARGIN = 2 (FENCE_TYPE = 8)
- AVOID_MARGIN = 9 (AVOID_ENABLE = 3)

In this configuration, I was hopping that the drone would stay at 9 m (= AVOID_MARGIN) of the ground but it only stop to descent at 4m (= FENCE_ALT_MIN + FENCE_MARGIN).

I still have some difficulties to understand how to debug using SITL so I don't have solution but I just wanted to triggered the problem yet. I may miss something with the parameters though.

[LAF_test.zip](https://github.com/ArduPilot/ardupilot/files/1433492/LAF_test.zip)

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 1]
```diff
@@ -420,7 +420,35 @@ bool AP_Proximity::get_upward_distance(uint8_t instance, float &distance) const
...
+}
+
+
+
+
 bool AP_Proximity::get_upward_distance(float &distance) const
 {
-    return get_upward_distance(primary_instance, distance);
+    for (uint8_t i=0; i<num_instances; i++) {
+        if (get_orientation(i) == ROTATION_PITCH_90) {
```
Comment (Katawann) [inline libraries/AP_Proximity/AP_Proximity.cpp]: get_orientation return 0: default or 1: Upside/Down ! Not the actual position of the sensor so the get_upward/downward_distance don't return anything

**Comment by peterbarker** [comment]
@amilcarlucas This is terribly conflicted now and 18 months old.

There's also some serious issues in terms of shape - the include of a Copter include in AC_PosControl, for example.

What are your intents with this PR?

**Comment by amilcarlucas** [comment]
I intend to find a workaround to that copter include hack at some point and re-submit this.
Do you have any ideas how I could remove the hack ?

**Comment by peterbarker** [comment]
@amilcarlucas I guess we can close this now?

**Comment by amilcarlucas** [comment]
This is outdated, some of it got merged with #15288 the rest migrated to #16261

**Pull request closed without merging**